In [63]:
from __future__ import unicode_literals

import pandas as pd
import numpy as np
from sklearn import preprocessing
from matplotlib import pyplot as plt
from sklearn.preprocessing import MinMaxScaler
from sklearn.preprocessing import LabelEncoder, OneHotEncoder
from sklearn.model_selection import train_test_split

# 1. Preparing data - 28 weeks of life of the banana plant
---

In [64]:
luminosity = pd.read_csv('../../../../data/raw/FincaPorvenir/Metereologico/28-weeks_January-12_August-03_2018/' \
                                 'Luz_January-12_August-03_2018.csv', )

In [65]:
print(luminosity.shape)
luminosity.head()

(9752, 2)


Fecha:  Luz (lux)
0  2018-01-12 00:17:28          1
1  2018-01-12 00:47:28          1
2  2018-01-12 01:17:27          1
3  2018-01-12 01:47:27          1
4  2018-01-12 02:17:28          1

## 1.1. We evaluate if this dataset has null type `NaN`
---

In [66]:
print(luminosity.isnull().any())
luminosity.isnull().values.any()

Fecha:       False
Luz (lux)    False
dtype: bool


False

Don't have null values

## 1.2. Selecting  relevant index columns features
---

Since the dataset has a column called **`Fecha:`**, which is not a numerical value,

it will be removed so that it does not interfere **with our subsequent scaling**, 

so we are only going to reference the values or samples of the column

** `Luz (lux)` ** and assigning them to the matrix 

`luminosity_luxes` created such as follow:


In [67]:
luminosity_luxes = luminosity.iloc[:, 1].values

print(type(luminosity_luxes))
# When we use iloc() dataframe function, the variable is turned on numpy array

luminosity_luxes

<class 'numpy.ndarray'>


array([1, 1, 1, ..., 1, 1, 1])

In [68]:
# Rehape the luminosity_luxes array 
luminosity_luxes = luminosity_luxes.reshape(-1,1)

luminosity_luxes

array([[1],
       [1],
       [1],
       ...,
       [1],
       [1],
       [1]])

## 1.3 Generating descriptive data to dataset
---


In [69]:
col = ['Luz (lux)']
luminosity_luxes_df = pd.DataFrame(luminosity_luxes, columns=col)
luminosity_luxes_df_describe = luminosity_luxes_df.describe()

In [70]:
luminosity_luxes_df_describe

Luz (lux)
count   9752.000000
mean   11228.888331
std    15119.881859
min        1.000000
25%        1.000000
50%      600.000000
75%    21674.000000
max    55731.000000

In [71]:
# https://pandas.pydata.org/pandas-docs/stable/generated/pandas.DataFrame.to_csv.html
# Export this descriptive data to comma separated values and java script object notation
luminosity_luxes_df_describe.to_csv('../../../../data/interim/Luminosity/28-weeks_January-12_August-03_2018/' +'\n' 
                              'Luminosity_Describe_January-12_August-03.cvs', sep=',', header=True, index=True)
luminosity_luxes_df_describe.to_json('../../../../data/interim/Luminosity/28-weeks_January-12_August-03_2018/' +'\n'
                                     'Luminosity_Describe_January-12_August-03.json')

## 2. Creating luminosity Training and Testing datasets

We have a **`luminosity_luxes_df`** dataset with 9752 samples rows.



In [72]:
print(luminosity_luxes_df.shape)
luminosity_luxes_df.head()

(9752, 1)


Luz (lux)
0          1
1          1
2          1
3          1
4          1

We'll divide it into two differents datasets:

- Training dataset
- Testing dataset

This is executed through [train_test_split](http://scikit-learn.org/stable/modules/generated/sklearn.model_selection.train_test_split.html  "sklearn.model_selection.train_test_split") function of this way:

`train_test_split` receive as a data parameter a numpy array, we have to turn the 

**`luminosity_luxes_df`** dataframe to numpy array such as follow:

In [73]:
# numpy_luminosity_luxes = luminosity_luxes.reset_index().values
numpy_luminosity_luxes = luminosity_luxes_df.values

In [74]:
# My numpy_luminosity_luxes variable now is a numpy array
print(numpy_luminosity_luxes.shape)
print(type(numpy_luminosity_luxes))
numpy_luminosity_luxes

(9752, 1)
<class 'numpy.ndarray'>


array([[1],
       [1],
       [1],
       ...,
       [1],
       [1],
       [1]])

We compose the following datasets from **`numpy_luminosity_luxes`** array :

- `luminosity_luxes_train`, which is the training matrix
- `luminosity_luxes_test`, the testing matrix

We use the [train_test_split](http://scikit-learn.org/stable/modules/generated/sklearn.model_selection.train_test_split.html  "sklearn.model_selection.train_test_split") function to create the training and testing dataset. Their parameters are:

![alt text](https://cldup.com/hukPWvvLxt-3000x3000.png "klearn.model_selection.train_test_split")


- The first parameter **should be an array**, then we pass the **`numpy_luminosity_luxes`** which contain all column features of luminosity.

Of this way, we pass all data (9752 samples rows) to from them `luminosity_luxes_train`

and `luminosity_luxes_test` will be created

- The `test_size=0.5` parameter means a 50% division; which means 
that half of the data goes to the test dataset and the other half 
goes to the training dataset.

A good choice for the size of tests **is usually 0.2 ie 20% or 0.25 or even 30%.** 

In some rare cases we will have 40% but almost never 0.5 or 50%

We choose 20% which means that we will have 20% of 9752 samples or observations for the test data set, 

in this case **9752 * 0.2 = 1950,4 samples or records for the test data set**


- The `train_size` parameter is the training dataset size.  **`test_size + train_size = 1 or 100%`**, 
then isn't necessary include it, because wheter we include to `test_size = 0.2`, then the remaining 
data will be to `train_size` this means **0.8 or 80%**

This means that **9752 * 0.8 = 7801,6  training dataset samples rows. **


- random_state is a seed or data source for generating random values for the data sets. 

If this parameter is not passed, the data will be generated in a random way, but in the way as by default numpy works them.

### 2.1  Creating Training and testing wind direction datasets
---

- `luminosity_luxes_train`, which is the training matrix
- `luminosity_luxes_test`, the testing matrix

In [75]:
luminosity_luxes_train, luminosity_luxes_test = train_test_split(numpy_luminosity_luxes, test_size = 0.2)

In [76]:
# We have 7801 rows to luminosity_luxes_train
print(type(luminosity_luxes_train))
print("The dimensionality of wind_direction training dataset is: " +'\n' , luminosity_luxes_train.shape)
print('\n')

# And we have 1951 rows to luminosity_luxes_test
print(type(luminosity_luxes_test))
print("The dimensionality of wind_direction testing dataset is: " +'\n' , luminosity_luxes_test.shape)

<class 'numpy.ndarray'>
The dimensionality of wind_direction training dataset is: 
 (7801, 1)


<class 'numpy.ndarray'>
The dimensionality of wind_direction testing dataset is: 
 (1951, 1)


In this way, the model, as you progressively learn the correlations in the training set, the better the prediction of the results in the test set.

But if the model learns too much from memory the correlations of the training sets, 

that is to say, when one learns from memory and does not understand things, then it 

will have problems to predict what is happening on the set of tests, because it is 

learned for difficult correlations, if the logic is not well understood and you can 

not make good predictions. This is called overfitting or overfitting


The really important thing is to understand that we need to have two different datasets

- Training set with which the ML model learns
- Test set, on which we test whether the ML model correctly learned the correlations

---
##  3. Feature Scaling Luminosity training and testing dataset  
---

[This article post](http://benalexkeen.com/feature-scaling-with-scikit-learn/  "Feature Scaling with scikit-learn") it's a great reference to explore the features scaling methods
on scikit learn

- `StandardScaler` assume that data is normally distributed at the level of each characteristic or variable. If the data is not normally distributed, it is not the best alternative to use for scaling.  

- `Min-Max Scaler` it is probably the most famous scaling algorithm and what it does is resize the range to leave it in a dimension of 0 to 1 or -1 and 1 (in case there are negative values in the original dataset of input data)

This scale of maximums and minimums works best for cases where standard scaling may not work properly. If the distribution is not Gaussian or the standard deviation is very small the escalation of maximums and minimums is the best idea.
However, it is sensitive to outliers or outliers, so if there are outliers in the data it is better to consider robust scaling.

- `Robust Scaler` it is similar to the previous one of maximums and minimums, only that it uses interquartile ranges instead of maximums and minimums, which makes it robust for the outliers

- `Normalizer` which scales each value, dividing each value by its magnitude in n dimensional spaces for n number of characteristics.


---



We will use the scaling of maximums and minimums to scale the water table data, because the standard deviation is very small, 

it does not have atypical values and it does not follow a normal distribution (you have to check this)

We apply the maximum and minimum scaling. We provide a rank or base scale that **will be between 0 and 1** using an object,

[MinMaxScaler](http://scikit-learn.org/stable/modules/generated/sklearn.preprocessing.MinMaxScaler.html "sklearn.preprocessing.MinMaxScaler") which transforms each characteristic, (in this 

case it will be the columns feature of `Luz (lux)`) individually according to a given range.

Product of its applicability, generates these attributes in the dataset, already transformed: 

![alt text](https://cldup.com/lTIv4HXgTk-3000x3000.png "sklearn.preprocessing.MinMaxScaler")

### 3.1 We apply maximium and minimum feature scaling to Luminosity training dataset

In [95]:
# We provide a base scale range
scaler = MinMaxScaler(feature_range=(0, 1))

print("Remember our luminosity training data " + '\n', luminosity_luxes_train)

Remember our luminosity training data 
 [[    1]
 [39330]
 [27756]
 ...
 [    1]
 [15820]
 [ 2503]]


With the [fit](http://scikit-learn.org/stable/modules/generated/sklearn.preprocessing.MinMaxScaler.html#sklearn.preprocessing.MinMaxScaler.fit  "MinMaxScaler.fit")
we compute the maximum and minimum value of  `luminosity_luxes_train` dataset to be used in the subsequent scaling 

We assing these values to `minmax_scale_training` variable.

In [84]:
minmax_scale_training = scaler.fit(luminosity_luxes_train.astype(float))
# print(minmax_scale_training.data_max_)
# http://terrapinssky.blogspot.com/2017/10/pythonresolved-dataconversionwarning.html

Then, we apply the [transform](http://scikit-learn.org/stable/modules/generated/sklearn.preprocessing.MinMaxScaler.html#sklearn.preprocessing.MinMaxScaler.transform "MinMaxScaler.transform") method to transform these data to maximum and mínimum scale value. 

Here, with this process, the `luminosity_luxes_train` data are scaled between **0 to 1**  selected range 


In [86]:
# transform luminosity_luxes_train data to maximum and mínimum scale value. 
luminosity_luxes_minmax_training = minmax_scale_training.transform(luminosity_luxes_train)

In [87]:
print ("And, these are our scaled data: " + '\n')
luminosity_luxes_minmax_training

And, these are our scaled data: 



array([[0.        ],
       [0.70570608],
       [0.4980262 ],
       ...,
       [0.        ],
       [0.28385071],
       [0.04489503]])

In [88]:
print('Luminosity Training dataset. Minimum value after MaxMinScaler:\nLuminosity={:.1f}'
      .format(luminosity_luxes_minmax_training[:,0].min()))

print('Luminosity Training dataset. Maximum value after MaxMinScaler:\nLuminosity={:.1f}'
      .format(luminosity_luxes_minmax_training[:,0].max()))

Luminosity Training dataset. Minimum value after MaxMinScaler:
Luminosity=0.0
Luminosity Training dataset. Maximum value after MaxMinScaler:
Luminosity=1.0


Then, our  MinMaxScaler normalized training dataset is `luminosity_luxes_minmax_training` numpy array

- We export this array to comma separated values 


In [90]:
luminosity_luxes_train_df = pd.DataFrame(luminosity_luxes_minmax_training, columns=col)

In [92]:
print(luminosity_luxes_train_df.shape)
luminosity_luxes_train_df.head()

(7801, 1)


Luz (lux)
0   0.000000
1   0.705706
2   0.498026
3   0.000000
4   0.000000

In [93]:
luminosity_luxes_train_df[luminosity_luxes_train_df['Luz (lux)']>1]

Luz (lux)
2671        1.0

In this way we have the dataset `luminosity_luxes_train_df` standardized and training, and export it to a .csv file

In [94]:
luminosity_luxes_train_df.to_csv('../../../../data/processed/Luminosity/28-weeks_January-12_August-03_2018/' +'\n' 
                                 'Luminosity_Normalized_TRAINING_January-12_August-03.csv', sep=',', header=True, index=False)

### 3.2 We apply maximium and minimum feature scaling to Luminosity testing dataset

In [96]:
# We provide a base scale range
scaler = MinMaxScaler(feature_range=(0, 1))

print("Remember our luminosity testing data " + '\n', luminosity_luxes_test)

Remember our luminosity testing data 
 [[    1]
 [22417]
 [    1]
 ...
 [    1]
 [    1]
 [ 8242]]


In [107]:
minmax_scale_test = scaler.fit(luminosity_luxes_test.astype(float))
# transform luminosity_luxes_test data to maximum and mínimum scale value. 
luminosity_luxes_minmax_test = minmax_scale_test.transform(luminosity_luxes_test)

In [101]:
print ("And, these are our testing scaled data: " + '\n')
luminosity_luxes_minmax_test

And, these are our testing scaled data: 



array([[0.        ],
       [0.42072862],
       [0.        ],
       ...,
       [0.        ],
       [0.        ],
       [0.15467633]])

In [102]:
print('Luminosity Testing dataset. Minimum value after MaxMinScaler:\nLuminosity={:.1f}'
      .format(luminosity_luxes_minmax_test[:,0].min()))

print('Luminosity Testing dataset. Maximum value after MaxMinScaler:\nLuminosity={:.1f}'
      .format(luminosity_luxes_minmax_test[:,0].max()))

Luminosity Testing dataset. Minimum value after MaxMinScaler:
Luminosity=0.0
Luminosity Testing dataset. Maximum value after MaxMinScaler:
Luminosity=1.0


Then, our  MinMaxScaler normalized training dataset is `luminosity_luxes_minmax_test` numpy array

- We export this array to comma separated values

In [103]:
luminosity_luxes_test_df = pd.DataFrame(luminosity_luxes_minmax_test, columns=col)

In [104]:
print(luminosity_luxes_test_df.shape)
luminosity_luxes_test_df.head()

(1951, 1)


Luz (lux)
0   0.000000
1   0.420729
2   0.000000
3   0.754969
4   0.000000

In [105]:
luminosity_luxes_test_df[luminosity_luxes_test_df['Luz (lux)']>1]

Empty DataFrame
Columns: [Luz (lux)]
Index: []

In this way we have the dataset luminosity_luxes_test_df standardized and testing, and export it to a .csv file

In [106]:
luminosity_luxes_train_df.to_csv('../../../../data/processed/Luminosity/28-weeks_January-12_August-03_2018/' +'\n' 
                                 'Luminosity_Normalized_TESTING_January-12_August-03.csv', sep=',', header=True, index=False)